File: Flutter/packages/tmk_translation_flutter/ios/Classes/TmkSettingsStore.swift
import Foundation

struct TmkPluginSettings {
    var diagnosisEnabled: Bool
    var consoleLogEnabled: Bool
    var networkEnvironment: String
    var mockEngineEnabled: Bool

    static let `default` = TmkPluginSettings(
        diagnosisEnabled: false,
        consoleLogEnabled: true,
        networkEnvironment: "test",
        mockEngineEnabled: false
    )

    init(diagnosisEnabled: Bool,
         consoleLogEnabled: Bool,
         networkEnvironment: String,
         mockEngineEnabled: Bool) {
        self.diagnosisEnabled = diagnosisEnabled
        self.consoleLogEnabled = consoleLogEnabled
        self.networkEnvironment = networkEnvironment
        self.mockEngineEnabled = mockEngineEnabled
    }

    init(map: [String: Any]) {
        let defaults = TmkPluginSettings.default
        self.diagnosisEnabled = map["diagnosisEnabled"] as? Bool ?? defaults.diagnosisEnabled
        self.consoleLogEnabled = map["consoleLogEnabled"] as? Bool ?? defaults.consoleLogEnabled
        self.networkEnvironment = map["networkEnvironment"] as? String ?? defaults.networkEnvironment
        self.mockEngineEnabled = map["mockEngineEnabled"] as? Bool ?? defaults.mockEngineEnabled
    }

    func toMap() -> [String: Any] {
        [
            "diagnosisEnabled": diagnosisEnabled,
            "consoleLogEnabled": consoleLogEnabled,
            "networkEnvironment": networkEnvironment,
            "mockEngineEnabled": mockEngineEnabled
        ]
    }
}

struct TmkSettingsStore {
    private enum Key {
        static let diagnosisEnabled = "tmk_translation_flutter.diagnosisEnabled"
        static let consoleLogEnabled = "tmk_translation_flutter.consoleLogEnabled"
        static let networkEnvironment = "tmk_translation_flutter.networkEnvironment"
        static let mockEngineEnabled = "tmk_translation_flutter.mockEngineEnabled"
        static let schemaVersion = "tmk_translation_flutter.schemaVersion"
    }

    private let userDefaults: UserDefaults

    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
    }

    func load() -> TmkPluginSettings {
        guard userDefaults.object(forKey: Key.schemaVersion) != nil else {
            return .default
        }
        return TmkPluginSettings(
            diagnosisEnabled: userDefaults.bool(forKey: Key.diagnosisEnabled),
            consoleLogEnabled: userDefaults.object(forKey: Key.consoleLogEnabled) as? Bool ?? true,
            networkEnvironment: userDefaults.string(forKey: Key.networkEnvironment) ?? "test",
            mockEngineEnabled: userDefaults.bool(forKey: Key.mockEngineEnabled)
        )
    }

    func save(_ settings: TmkPluginSettings) {
        userDefaults.set(settings.diagnosisEnabled, forKey: Key.diagnosisEnabled)
        userDefaults.set(settings.consoleLogEnabled, forKey: Key.consoleLogEnabled)
        userDefaults.set(settings.networkEnvironment, forKey: Key.networkEnvironment)
        userDefaults.set(settings.mockEngineEnabled, forKey: Key.mockEngineEnabled)
        userDefaults.set(1, forKey: Key.schemaVersion)
    }
}

File: Flutter/packages/tmk_translation_flutter/android/src/main/kotlin/co/timekettle/translation/flutter/TmkSettingsStore.kt
package co.timekettle.translation.flutter

import android.content.Context

internal data class TmkSettings(
    val diagnosisEnabled: Boolean,
    val consoleLogEnabled: Boolean,
    val networkEnvironment: String,
    val mockEngineEnabled: Boolean,
) {
    fun toMap(): Map<String, Any> {
        return mapOf(
            "diagnosisEnabled" to diagnosisEnabled,
            "consoleLogEnabled" to consoleLogEnabled,
            "networkEnvironment" to networkEnvironment,
            "mockEngineEnabled" to mockEngineEnabled,
        )
    }

    companion object {
        fun defaults(): TmkSettings {
            return TmkSettings(
                diagnosisEnabled = false,
                consoleLogEnabled = true,
                networkEnvironment = "test",
                mockEngineEnabled = false,
            )
        }

        fun fromMap(raw: Map<*, *>?): TmkSettings {
            val defaults = defaults()
            return TmkSettings(
                diagnosisEnabled = raw?.get("diagnosisEnabled") as? Boolean ?: defaults.diagnosisEnabled,
                consoleLogEnabled = raw?.get("consoleLogEnabled") as? Boolean ?: defaults.consoleLogEnabled,
                networkEnvironment = raw?.get("networkEnvironment") as? String ?: defaults.networkEnvironment,
                mockEngineEnabled = raw?.get("mockEngineEnabled") as? Boolean ?: defaults.mockEngineEnabled,
            )
        }
    }
}

internal class TmkSettingsStore(context: Context) {
    private val preferences = context.getSharedPreferences("tmk_translation_flutter", Context.MODE_PRIVATE)

    fun load(): TmkSettings {
        if (!preferences.contains(KEY_NETWORK_ENVIRONMENT)) {
            return TmkSettings.defaults()
        }
        return TmkSettings(
            diagnosisEnabled = preferences.getBoolean(KEY_DIAGNOSIS_ENABLED, false),
            consoleLogEnabled = preferences.getBoolean(KEY_CONSOLE_LOG_ENABLED, true),
            networkEnvironment = preferences.getString(KEY_NETWORK_ENVIRONMENT, "test") ?: "test",
            mockEngineEnabled = preferences.getBoolean(KEY_MOCK_ENGINE_ENABLED, false),
        )
    }

    fun save(settings: TmkSettings) {
        preferences.edit()
            .putBoolean(KEY_DIAGNOSIS_ENABLED, settings.diagnosisEnabled)
            .putBoolean(KEY_CONSOLE_LOG_ENABLED, settings.consoleLogEnabled)
            .putString(KEY_NETWORK_ENVIRONMENT, settings.networkEnvironment)
            .putBoolean(KEY_MOCK_ENGINE_ENABLED, settings.mockEngineEnabled)
            .apply()
    }

    private companion object {
        private const val KEY_DIAGNOSIS_ENABLED = "diagnosis_enabled"
        private const val KEY_CONSOLE_LOG_ENABLED = "console_log_enabled"
        private const val KEY_NETWORK_ENVIRONMENT = "network_environment"
        private const val KEY_MOCK_ENGINE_ENABLED = "mock_engine_enabled"
    }
}
